k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapConnection.java

```diff
@@ -356,6 +368,46 @@ private void authenticate() throws MessagingException, IOException {
         }
     }
 
+    private void authXoauth2withSASLIR() throws IOException, MessagingException {
+        try {
+            attemptXOAuth2();
+        } catch (NegativeImapResponseException e) {
+            //We could avoid this if we had a reasonable chance of knowing
+            //if a token was invalid before use (e.g. due to expiry). But we don't
+            //This is the intended behaviour per AccountManager
+            Log.v(LOG_TAG, "Authentication exception, invalidating token and re-trying", e);
```

Comment:
It's possible they are just HTTP Status codes - the SMTP example I added from Google's documentation has 401 (Unauthorised). Maybe I should not retry on 5xx error codes.